만국박람회 [STEP 1] Mary, Kobe

devKobe24 opened the pull request:
# 만국박람회 [STEP 1]

# STEP 1 PR

안녕하세요 제임스(@inwoodev)!
만국박람회 STEP 1 PR 보내드립니다.
설레이는 제임스와의 첫 PR입니다. 🙌
첫 리뷰 잘 부탁드립니다!🙇‍♀️


## 고민했던 점
1️⃣ DTO type `Decodable` 프로토콜 채택.
Codable과 Decodable 중 어떤 프로토콜을 채택할지 고민이 많았습니다.
결과적으로 현재 프로젝트에서는 서버에 따로 올려줄 데이터가 없다고 판단하여 Decodable을 채택하게 되었습니다.

2️⃣ `CodingKey`을 이용한 json타입 키 매칭.
json의 키 이름이 swift의 네이밍 컨벤션과 달라서 고민했습니다.
따라서 이를 맞춰주기위하여 `CodingKeys` 열거형을 생성, case에 원시값으로 네이밍 컨벤션을 맞춰 주었습니다.

3️⃣ `description` 프로퍼티명 수정.
`CodingKey`프로토콜이 `CustomStringConvertible`을 채택하기 때문에 `description`이라는 이름의 프로퍼티가 이미 선언되어 있었습니다. 따라서 `description`이라는 프로퍼티 명을 사용할 수 없는 상황에 처해 많은 고민 끝에 그대로 `desc` 프로퍼티명을 사용하기로 하였습니다.

4️⃣ Unit Test 작성.
json 파일이 파싱이 잘 되는지 확인하고 싶었고 그 결과를 어떻게 확인하여야 할 지 고민했습니다.
그에 따라 테스트 코드를 작성하였습니다.
결과적으로 파싱이 잘되어 데이터가 잘 들어오는 것을 확인했습니다.

## 조언을 얻고 싶은 부분
🙋‍♂️ 테스트 코드에서 초기화 해줄 인스턴스가 없어 setUpWithError() 메서드와 tearDownWithError()메서드를 삭제하였습니다. 이 두 메서드를 삭제해도 괜찮은지 궁금합니다.

MaryJo-github committed:
:sparkles: Feat: ExpoModel 생성 및 구현

Commit:
:sparkles: Feat: ItemsModel 생성 및 구현.

출품작 상세화면을 위한 모델을 구현하였습니다.

MaryJo-github committed:
:recycle: Refactor: ExpoModel 변수를 상수로 변경

Commit:
:recycle: Refactor: ItemsModel 코드 수정, ExpoModel 코드 수정.

ItemsModel이  Decodable 프로토콜을 채택했습니다, CodingKeys의 case를 추가했습니다. ExpoModel이 Decodable 프로토콜을 채택했습니다.

MaryJo-github committed:
:white_check_mark: Test: ExpoModel unit test 추가

Commit:
:fire: Remove: 불필요한 테스트 코드 메서드 삭제.

ExpoModelTests의 setUpWithError, tearDownWithError을 삭제하였습니다.

Commit:
:recycle: Refactor: ItemsModel 프로퍼티명 변경.

ItemsModel이 채택하고 있는  CodingKey가 description 프로퍼티를 선언하고 있으므로 ItemsModel 프로퍼티명을 다시 수정하였습니다.

Commit:
:white_check_mark: Test: ItemsModelTests 생성 및 구현.

MaryJo-github committed:
:recycle: Refactor: 잘못된 변수명 수정.

Test case의 잘못된 변수명을 수정하였습니다.

inwoodev commented on Expo1900/ExpoModelTests/ExpoModelTests.swift:
```diff
@@ -0,0 +1,34 @@
+import XCTest
+@testable import Expo1900
+
+final class ExpoModelTests: XCTestCase {
+    func test_파리만국박람회json데이터를_parsing했을때_정보를추출한다() {
+        // given
+        guard let dataAsset = NSDataAsset(name: "exposition_universelle_1900") else {
+            return
+        }
+        let decoder = JSONDecoder()
+        let titleExpectation = "파리 만국박람회 1900(L\'Exposition de Paris 1900)"
+        let visitorExpectation = 48130300
+        let locationExpectation = "프랑스 파리"
+        let durationExpectation = "1900. 04. 14 - 1900. 11. 12"
+        let descriptionExpectation = "1900년 파리 만국박람회(지금의 세계 박람회[World’s Fair/Exposition/Expo])는 지난 세기를 기념하고 다음 세기를 향한 발전을 가속하자는 의미에서 1900년 4월 14일부터 11월 12일까지 프랑스 파리에서 열린 세계 박람회였다. 이 박람회에서 널리 선보였던 건축 양식이 바로 '아르누보'였다. 총 관람객수가 5000만 명에 달한 1900년 세계 박람회에서는 수많은 기계와 발명품, 그리고 건축물들이 전시됐는데 그 중에는 그랑드 루 드 파리 대관람차, 마트료시카 인형, 디젤 엔진, 유성영화, 에스컬레이터, 텔레그라폰 (최초의 자석식 녹음기) 등 지금도 널리 알려져 있는 것들도 등장했다.\n\n프랑스는 1855년에도 만국 ...
+        
+        // when
+        let expo = try! decoder.decode(ExpoModel.self, from: dataAsset.data)
+        
+        // then
+        XCTAssertEqual(expo.title, titleExpectation)
+        XCTAssertEqual(expo.visitors, visitorExpectation)
+        XCTAssertEqual(expo.location, locationExpectation)
+        XCTAssertEqual(expo.duration, durationExpectation)
+        XCTAssertEqual(expo.description, descriptionExpectation)
+    }
```
개인적으로 구현사항에 있어서 유닛테스트를 통해 행동 검증을 하는 것은 아주 좋은 습관이라고 생각합니다.👍👍

유닛테스트의 필요성을 어떻게 느끼셨는지 궁금합니다!

devKobe24 replied:
#### 코멘트에 대한 코비와 메리의 답변 🔑
- 답변
    - 저희는 JSON 데이터를 서버가 넘겨준다는 가정을 하였습니다. 그래서 서버가 넘겨주기 전 상황을 상상 해봤습니다. 서버에서 데이터가 넘어오기 전 프론트에서 할 수 있는 일은 먼저 벡엔드와 상의 후에 구조가 이미 짜여져 있는 상황이니 미리 구조를 짜놓는 일이라고 생각했습니다. 그러면 미리 구조를 짜놓은 뒤 이 구조가 제대로 짜였는지 검증하기 위해서는 데이터가 제대로 넘어오는지 확인하는 일이라고 생각했습니다. 그로인해 유닛테스트의 필요성을 느끼게 되었습니다.

inwoodev commented on Expo1900/ExpoModelTests/ExpoModelTests.swift:
```diff
@@ -0,0 +1,34 @@
+        let descriptionExpectation = "1900년 파리 만국박람회(지금의 세계 박람회[World’s Fair/Exposition/Expo])는 지난 세기를 기념하고 다음 세기를 향한 발전을 가속하자는 의미에서 1900년 4월 14일부터 11월 12일까지 프랑스 파리에서 열린 세계 박람회였다. 이 박람회에서 널리 선보였던 건축 양식이 바로 '아르누보'였다. 총 관람객수가 5000만 명에 달한 1900년 세계 박람회에서는 수많은 기계와 발명품, 그리고 건축물들이 전시됐는데 그 중에는 그랑드 루 드 파리 대관람차, 마트료시카 인형, 디젤 엔진, 유성영화, 에스컬레이터, 텔레그라폰 (최초의 자석식 녹음기) 등 지금도 널리 알려져 있는 것들도 등장했다.\n\n프랑스는 1855년에도 만국 ...
+        
+        // when
+        let expo = try! decoder.decode(ExpoModel.self, from: dataAsset.data)
```
유닛테스트를 하여 한 발자국 더 나아간 것은 매우 바람직하나 요구사항을 지키는 것이 저는 개인적으로 더 중요하다고 생각합니다. 
개인적으로 저는 적절히 주석과 느낌표를 사용하신것 같지만 이번 프로젝트 제약사항을 어긴 것이기에 조금 더 마이너스요소로 작용할 것 같습니다.

devKobe24 replied:
#### 코멘트에 대한 코비와 메리의 답변 🔑
- 답변 1
    - 다음과 같은 근거로서 주석과 느낌표를 사용하였습니다. 주석의 경우 give, when, then을 구분하기 위하여 사용하였습니다. 이렇게 구분지어 확실하게 코드를 작성하는 습관이 좋다는 잼킹의 말씀이 있으셨습니다. 그리고 느낌표 또한 유닛테스트의 조건 중 빠른 테스트 구현을 위하여 느낌표를 사용하여도 좋다는 잼킹의 활동학습 시간의 이야기가 있었어서 사용하였습니다. 
- 답변 2
    - 테스트 케이스는 기본 로직 코드와 달리 given when then을 주석으로 명시하면 테스트를 파악하기 더 좋을 것 같아 추가하였습니다. 또한 성공하는 케이스에서 강제추출을 사용했을 때 옵셔널바인딩에 실패하여 에러가 발생한다면 그것대로 에러를 잡아낸 것이라고 생각하여 사용해도 괜찮다고 생각했습니다. 그래도 제임스의 말씀대로 테스트케이스에서도 주석과 강제추출을 사용하지 않는 것이 좋을까요?

inwoodev replied:
흠 이 부분은 관점의 차이라고 볼 수 있겠네요. 요구사항의 범위를 로직 코드로 한정지으신것 같아요! 음 저라면 요구사항에서 이렇게 모호한 부분을 적극적으로 조율한 뒤 도출되는 결과를 기반으로 작업에 임할 것 같아요~ 나중에 기업 사전과제를 하실 때 요구사항에서 애매한 부분이나 모호한 부분을 적극적으로 물어보는 것이 플러스 요인으로 작용할 수 있다고 생각합니다 :)

inwoodev commented on Expo1900/expo_assets/bangjja.imageset/bangjja~universal@1x.png:
이번 프로젝트 요구사항에서 사진 추가는 없는 것 같습니다~

devKobe24 replied:
#### 코멘트에 대한 코비와 메리의 답변 🔑
- 답변
    - 에셋 데이터를 프로젝트 파일 내부에 넣어 둬야겠다는 판단을 했습니다. 그 이유는 데이터 관리 측면에서 편리할 것이라는 판단을 했습니다. 하지만 제임스의 말씀대로 프로젝트 요구사항에는 없는 사항이여서 유닛테스트를 위한 JSON 파일 외에는 삭제하는 것이 옳은 판단으로 생각되어 삭제하겠습니다.

inwoodev commented on Expo1900/Expo1900/ItemsModel.swift:
```diff
@@ -0,0 +1,19 @@
+//  Created by MARY, KOBE on 2023/06/27.
+//
+
+struct ItemsModel: Decodable {
+	let name: String
+	let imageName: String
+	let shortDescription: String
+	let desc: String
+	
+	enum CodingKeys: String, CodingKey {
+		case name, desc
+		case imageName = "image_name"
+		case shortDescription = "short_desc"
+	}
+}
```
깔끔합니다~👍👍

CodingKey는 왜 필요할까요~? 두 분이 공부하신 내용을 저도 공유받고 싶습니다 :)

그리고 선언하신 `CodingKeys`를 외부에서 접근할 필요가 있는지도 고민해 볼 필요가 있어보입니다~

devKobe24 replied:
#### 코멘트에 대한 코비와 메리의 답변 🔑
- 답변
    - `CodingKey`가 필요한 이유는 일정한 네이밍 컨벤션을 유지하기 위해서 입니다. 그로인해 가독성이 좋아집니다. 코드를 전달받아 해석하는 사람의 입장에서 흐름이 끊기지 않고 해석되게 합니다. 
    - `CodingKey`를 외부에서 접근할 필요는 없어보입니다. `private` 접근 제한자를 사용하는 것으로 수정하겠습니다.

MaryJo-github replied:
**추가 답변** 🔑
제임스가 공유해주신 공식문서를 통해 JSONDecoder의 keyDecodingStrategy을 알게되었습니다.
이를 이용하면 snake case를 camel case로 간편하게 변환할 수 있어서 CodingKey 대신 해당 방법으로 수정하였습니다.

inwoodev commented on Expo1900/Expo1900/ExpoModel.swift:
```diff
@@ -0,0 +1,14 @@
+//  Created by MARY, KOBE on 2023/06/27.
+//
+
+struct ExpoModel: Decodable {
+    let title: String
+    let visitors: Int
+    let location: String
+    let duration: String
+    let description: String
+}
```
👍👍👍👍

혹시 Codable, Decodable, 그리고 Encodable의 차이점을 공부하셨을까요? 관련된 내용을 공유해주시면 감사하겠습니다!

그리고 추가로 해당 모델의 프로퍼티의 접근레벨이 어떻게 되는지 아시는지 그리고 아신다면 왜 해당 접근레벨로 프로퍼티들을 선언하셨는지 궁금합니다~

devKobe24 replied:
#### 코멘트에 대한 코비와 메리의 답변 🔑
- 답변 1
    - `Encodable`은 Swift의 내장 프로토콜 중 하나로, 내부 표현 (즉, Swift 객체)을 외부 표현 (예: JSON)으로 변환하는 방법을 정의합니다. 다시말해, 자신을 외부 표현으로 인코딩할 수 있는 형식입니다.
    - `Decodable`은 Swift의 내장 프로토콜 중 하나로, JSON과 같은 외부 표현을 내부 표현 (즉, Swift 객체)으로 변환하는 방법을 정의합니다. 다시말해, 외부 표현에서 자신을 디코딩할 수 있는 형식입니다.
    - `Codable`은 Decodable과 Encodable 프로토콜을 합친 타입입니다. 주로 JSON 또는 유사한 데이터 형식을 Swift 객체로 변환(디코딩)하거나, Swift 객체를 JSON 형식으로 변환(인코딩)하는데 사용됩니다.
- 답변 2
    - 해당 모델의 프로퍼티 접근 레벨은 `internal`입니다. `internal`로 프로퍼티들을 선언한 이유는 `ExpoModel`의 모든 프로퍼티들은 같은 모듈(`Expo1900`) 내의 다른 코드에서 접근할 수 있지만, 다른 모듈에서는 접근할 수 없어야하기 때문입니다.

inwoodev replied:
답변이 아주 명료하네요 👍👍👍. 음...한가지만 더 여쭤보자면...Codable은 그럼 타입라고 부르는게 맞을까요?? 😉

inwoodev commented on Expo1900/Expo1900/ExpoModel.swift:
```diff
@@ -0,0 +1,14 @@
+
+struct ExpoModel: Decodable {
+    let title: String
+    let visitors: Int
```
방문객의 숫자가 음수가 될 수도 있을까요~ 🤔

devKobe24 replied:
#### 코멘트에 대한 코비와 메리의 답변 🔑
- 답변
    - 방문객의 숫자는 항상 양수일 것 같아 UInt로 변경해주었습니다!

inwoodev reviewed (commented):
안녕하세요~ 이번 리뷰를 맡게된 제임스입니다. 저도 부족한 점이 많지만 그래도 두 분이 성장하는데에 있어서 조금이나마 도움이 될 수 있도록 열심히 리뷰할게요 :)

### 인상적이었던 부분:
- 유닛테스트를 통해 구현사항을 한 번더 검증하는 과정을 거치신것이 매우 인상적이었어요!
- 고민하신 부분을 공유하며 근거있는 코드를 작성하시는 부분이 매우 인상적이에요!

### 고려하셨으면 하는 부분:
- 요구사항을 조금 더 꼼꼼하게 봐주시면 더 군더더기 없는 코드를 작성하실 수 있을 것 같아요!! 💪💪💪


## 고민하신 부분에 대한 코멘트:
> 1️⃣ DTO type Decodable 프로토콜 채택.
Codable과 Decodable 중 어떤 프로토콜을 채택할지 고민이 많았습니다.
결과적으로 현재 프로젝트에서는 서버에 따로 올려줄 데이터가 없다고 판단하여 Decodable을 채택하게 되었습니다.

꼭 필요한 부분이 무엇인지 파악하고 구현하신 부분은 정말로 칭찬 드리고 싶어요! 저도 남겨주신 근거에 의해서 설득되었습니다 :)

---

> 2️⃣ CodingKey을 이용한 json타입 키 매칭.
json의 키 이름이 swift의 네이밍 컨벤션과 달라서 고민했습니다.
따라서 이를 맞춰주기위하여 CodingKeys 열거형을 생성, case에 원시값으로 네이밍 컨벤션을 맞춰 주었습니다.

[공식문서](https://developer.apple.com/documentation/foundation/jsondecoder/keydecodingstrategy)도 한 번 보시면 다른 접근방식 또한 파악하실 수 있을 거에요! 

---

> 3️⃣ description 프로퍼티명 수정.
CodingKey프로토콜이 CustomStringConvertible을 채택하기 때문에 description이라는 이름의 프로퍼티가 이미 선언되어 있었습니다. 따라서 description이라는 프로퍼티 명을 사용할 수 없는 상황에 처해 많은 고민 끝에 그대로 desc 프로퍼티명을 사용하기로 하였습니다.

`longDescription`과 같은 네이밍 활용하여 의미가 더 명확해질 수도 있겠네요. 이건 저만 겪은 특이케이스일 수도 있지만 body가 엄청 길어질 때 조금이나마 용량을 줄이기 위해 아예 네이밍을 엄청 짧게 선언하는 경우도 있습니다 :)

---

> 4️⃣ Unit Test 작성.
json 파일이 파싱이 잘 되는지 확인하고 싶었고 그 결과를 어떻게 확인하여야 할 지 고민했습니다.
그에 따라 테스트 코드를 작성하였습니다.
결과적으로 파싱이 잘되어 데이터가 잘 들어오는 것을 확인했습니다.

훌륭합니다~ 유닛테스트를 작성하는 것도 좋지만 제가 생각할 때에 더 중요한 부분은 유닛테스트를 **견고하게** 작성하는 것이라고 생각해요~ 이 부분은 차차 더욱 디벨롭 해 가면 좋을 것 같아요. 음 만약 시간이 되신다면 실패하는 경우를 테스트하는 것도 유의미하다고 개인적으로 생각해요 :)

---

## 조언을 드리고 싶은 부분:

> 🙋‍♂️ 테스트 코드에서 초기화 해줄 인스턴스가 없어 setUpWithError() 메서드와 tearDownWithError()메서드를 삭제하였습니다. 이 두 메서드를 삭제해도 괜찮은지 궁금합니다.

해당 메서드의 용도를 잘 알고 계신듯하고 현 시점에서 필요없다는 부분도 인지하시기에 저도 삭제해도 괜찮다고 생각합니다!

---

고생하셨습니다~ 앞으로 남은 스텝이 더 기대가 되네요. 코멘트 한 번 보시고 궁금한 점이나 이해가 잘 안되는 부분이 있으면 언제든 글 남겨주세요~ 더 수정사항이 필요 없다 싶으면 머지하도록 할게요!

Commit:
:fire: Remove: 불필요한 Assets 파일 삭제.

요구사항에 맞지 않는 Assets 파일을 삭제하였습니다.

MaryJo-github committed:
:recycle: Refactor: ItemsModel의 CodingKey 삭제.

ItemsModel의 CodingKey를 삭제하고, KeyDecodingStrategy의convertFromSnakeCase로 수정하였습니다.

Commit:
:recycle: Refactor: ExpoModel의 visitors 타입 변경.

ExpoModel의 visitors 프로퍼티의 타입을 Int에서 UInt 타입으로 변환하였습니다.
